### PR TITLE
support least connections

### PR DESCRIPTION
**Pull request by jiuker**
support least connections

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -195,6 +195,8 @@ type BackendStats struct {
 	Tx              int64
 	UpSince         time.Time
 	DowntimeStart   time.Time
+	LastFinished    time.Time
+	CurrentCalls    int
```
Comment (klauspost) [inline main.go:199]: I think it is reasonable to make these atomic values. Use `int64` for both and `time.Time.UnixNano()` for timestamp.

If there are two updating the `LastFinished` at the same time it shouldn't really be a big issue.

File: main.go [hunk 2]
```diff
@@ -450,15 +452,37 @@ func (s *site) nextProxy() *Backend {
...
+			if backend.Stats.LastFinished.Before(earliest) {
+				earliest = backend.Stats.LastFinished
+				idx = i
+			}
+		}
+		backend.Stats.RUnlock()
+	}
 	// random backend from a list of available backends.
-	return backends[idx]
+	backend := backends[idx]
```
Comment (klauspost) [inline main.go:470]: Comment above no longer reflects code.

File: main.go [hunk 3]
```diff
@@ -450,15 +452,37 @@ func (s *site) nextProxy() *Backend {
...
+	backend.Stats.Lock()
+	backend.Stats.CurrentCalls++
+	backend.Stats.Unlock()
+	return backend
```
Comment (klauspost) [inline main.go:474]: Return a `func()` that cancels this.

```suggestion
	return backend, func() {
		backend.Stats.CurrentCalls--
		backend.Stats.LastFinished = time.Now()	
	}
```
(with either locking or atomics)

File: main.go [hunk 4]
```diff
@@ -450,15 +452,37 @@ func (s *site) nextProxy() *Backend {
...
 }
 
 // ServeHTTP - LoadBalancer implements http.Handler
 func (s *site) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	backend := s.nextProxy()
+	defer func() {
```
Comment (klauspost) [inline main.go:480]: `nil` backend must be handled. The deferred function is before that is checked.

```
backend, done := s.nextProxy()
if backend == nil || !backend.Online() {...}
defer done()
```

I don't really know why we return offline backends in the first place.

File: main.go [hunk 5]
```diff
@@ -764,7 +788,8 @@ func configureSite(ctx *cli.Context, siteNum int, siteStrs []string, healthCheck
 			Transport:      transport,
 			ModifyResponse: modifyResponse(),
 		}
-		stats := BackendStats{MinLatency: 24 * time.Hour, MaxLatency: 0}
+		off := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(len(endpoints))
+		stats := BackendStats{MinLatency: 24 * time.Hour, MaxLatency: 0, LastFinished: time.Now().Add(time.Duration(i + off%len(endpoints)))}
```
Comment (klauspost) [inline main.go:792]: What is the point of this?

**Comment by klauspost** [comment]
This is fixing #16 - right?

**Comment by jiuker** [comment]
> This is fixing #16 - right?
> 
> @harshavardhana While I do believe this is a better default than round-robin, I think RR should still be an option, or did you already discuss this?

yeah. Fixing #16

File: main.go (hunk 4 above)
Comment (jiuker) [reply main.go]: yeah. now is fine? @klauspost

**Comment by harshavardhana** [comment]
> @harshavardhana While I do believe this is a better default than round-robin, I think RR should still be an option, or did you already discuss this?

It can be an option, we can take a flag that says `--host-balance=leastconn` or `--host-balance=random` 

Currently what we support is random not round-robin

File: main.go [hunk 6]
```diff
@@ -183,6 +183,8 @@ func (b *Backend) getServerStatus() string {
 
 // BackendStats holds server stats for backend
 type BackendStats struct {
+	LastFinished int64
+	CurrentCalls int64
```
Comment (harshavardhana) [inline main.go:187]:
```suggestion
	LastFinished atomic.Int64
	CurrentCalls atomic.Int64
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: main.go (hunk 4 above)
Comment (klauspost) [reply main.go]: Looks great!

**Comment by klauspost** [comment]
Code LGTM, but some minor things:

A) change "leastconn" to just `least`
B) Default should be `least` - it is by far better overall. Deals better with requests of varying latency and varying host speed.
C) We should never select hosts that are marked offline, since this will return many more errors than needed.

File: main.go [hunk 7]
```diff
@@ -1026,6 +1042,9 @@ EXAMPLES:
 
   6. Sidekick as TLS terminator:
      $ sidekick --cert public.crt --key private.key --health-path=/minio/health/cluster http://site1-minio{1...4}:9000
+
+  7. Load balance across 4 MinIO Servers (http://minio1:9000 to http://minio4:9000), Set host balance as leastconn 
+     $ sidekick ----host-balance=leastconn "/minio/health/cluster" http://minio{1...4}:9000
```
Comment (harshavardhana) [inline main.go:1047]:
```suggestion
     $ sidekick --host-balance=leastconn --health-path=/minio/health/cluster http://minio{1...4}:9000
```

**Comment by jiuker** [comment]
> it is by far better overall

1 and 2 are there.
3. for this:
```go
func (s *site) upBackends() []*Backend {
	var backends []*Backend
	for _, backend := range s.backends {
		if backend.Online() {
			backends = append(backends, backend)
		}
	}
	return backends
}

func (s *site) nextProxy() (*Backend, func()) {
	backends := s.upBackends()
	if len(backends) == 0 {
		return nil, func() {}
	}
```
@klauspost

**Review by klauspost — APPROVED** [review]

File: main.go [hunk 8]
```diff
@@ -973,6 +1004,10 @@ func main() {
 			Name:  "status-code",
 			Usage: "filter by given status code",
 		},
+		cli.StringFlag{
+			Name:  "host-balance",
+			Usage: "set host balance as least or random",
```
Comment (harshavardhana) [inline main.go:1009]:
```suggestion
			Usage: "Specify the algorithm used to select a server when doing load balancing, supported values are 'least', 'random'",
			Value: "least",
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**